osu.Game/Beatmaps/Formats/LegacyBeatmapEncoder.cs

```diff
@@ -0,0 +1,399 @@
+            //     writer.WriteLine(@"AlwaysShowPlayfield: 1");
+            // if (b.OverlayPosition != OverlayPosition.NoChange)
+            //     writer.WriteLine(@"OverlayPosition: " + b.OverlayPosition);
+            // if (!string.IsNullOrEmpty(b.SkinPreference))
+            //     writer.WriteLine(@"SkinPreference:" + b.SkinPreference);
+            // if (b.EpilepsyWarning)
+            //     writer.WriteLine(@"EpilepsyWarning: 1");
+            // if (b.CountdownOffset > 0)
+            //     writer.WriteLine(@"CountdownOffset: " + b.CountdownOffset.ToString());
+            if (beatmap.BeatmapInfo.RulesetID == 3)
+                writer.WriteLine(FormattableString.Invariant($"SpecialStyle: {(beatmap.BeatmapInfo.SpecialStyle ? "1" : "0")}"));
+            writer.WriteLine(FormattableString.Invariant($"WidescreenStoryboard: {(beatmap.BeatmapInfo.WidescreenStoryboard ? "1" : "0")}"));
+            // if (b.SamplesMatchPlaybackRate)
+            //     writer.WriteLine(@"SamplesMatchPlaybackRate: 1");
+        }
+
+        private void handleEditor(TextWriter writer)
+        {
+            writer.WriteLine("[Editor]");
+
+            if (beatmap.BeatmapInfo.Bookmarks.Length > 0)
+                writer.WriteLine(FormattableString.Invariant($"Bookmarks: {string.Join(',', beatmap.BeatmapInfo.Bookmarks)}"));
+            writer.WriteLine(FormattableString.Invariant($"DistanceSpacing: {beatmap.BeatmapInfo.DistanceSpacing}"));
+            writer.WriteLine(FormattableString.Invariant($"BeatDivisor: {beatmap.BeatmapInfo.BeatDivisor}"));
+            writer.WriteLine(FormattableString.Invariant($"GridSize: {beatmap.BeatmapInfo.GridSize}"));
+            writer.WriteLine(FormattableString.Invariant($"TimelineZoom: {beatmap.BeatmapInfo.TimelineZoom}"));
+        }
+
+        private void handleMetadata(TextWriter writer)
+        {
+            writer.WriteLine("[Metadata]");
+
+            writer.WriteLine(FormattableString.Invariant($"Title: {beatmap.Metadata.Title}"));
+            writer.WriteLine(FormattableString.Invariant($"TitleUnicode: {beatmap.Metadata.TitleUnicode}"));
+            writer.WriteLine(FormattableString.Invariant($"Artist: {beatmap.Metadata.Artist}"));
+            writer.WriteLine(FormattableString.Invariant($"ArtistUnicode: {beatmap.Metadata.ArtistUnicode}"));
+            writer.WriteLine(FormattableString.Invariant($"Creator: {beatmap.Metadata.AuthorString}"));
+            writer.WriteLine(FormattableString.Invariant($"Version: {beatmap.BeatmapInfo.Version}"));
+            writer.WriteLine(FormattableString.Invariant($"Source: {beatmap.Metadata.Source}"));
+            writer.WriteLine(FormattableString.Invariant($"Tags: {beatmap.Metadata.Tags}"));
+            writer.WriteLine(FormattableString.Invariant($"BeatmapID: {beatmap.BeatmapInfo.OnlineBeatmapID ?? 0}"));
+            writer.WriteLine(FormattableString.Invariant($"BeatmapSetID: {beatmap.BeatmapInfo.BeatmapSet.OnlineBeatmapSetID ?? 0}"));
```

Comment:
Not sure about stable interop, but on the eye test this doesn't seem to preserve nulls (encoding and decoding would as far I can tell silently coalesce these `null`s to 0).

Reply:
Coalescing is fine, although OnlineBeatmapSetID should coalesce to -1 instead.